BAU: Make session store usage more general

andy-paine opened the pull request:
- I would recommend reviewing this on a per-commit basis
- Interesting things to note:
 * The tests in TransitionCacheTests (you should convince yourself that this is the correct behaviour for a read/write + write only cache)
 * This is essentially a no-op for the existing system (tests have only changed to handle new config format)
- The context here is that we would like to be able to use a second cache, either to support the transition to another serialization format or to another cache provider

Test diff: `2 files changed, 520 insertions(+), 8 deletions(-)` 🎉 
Non-test diff: ` 16 files changed, 394 insertions(+), 75 deletions(-)`

andy-paine commented on hub/policy/src/test/java/uk/gov/ida/hub/policy/controllogic/TransitionCacheTest.java:
```diff
@@ -0,0 +1,407 @@
+            put(new SessionId("key1"), "value1");
+            put(new SessionId("key2"), "value2");
+        }};
+        ConcurrentMap transition = new ConcurrentHashMap() {{
```
It's a test

andy-paine commented on hub/policy/src/main/java/uk/gov/ida/hub/policy/PolicyModule.java:
```diff
@@ -124,14 +124,16 @@ public JsonClient forSamlSoapProxy(JsonResponseProcessor responseProcessor, Poli
     @Singleton
-    public ConcurrentMap<SessionId, DateTime> datetime_cache(InfinispanCacheManager infinispanCacheManager) {
-        return infinispanCacheManager.getCache("datetime_cache");
+    public ConcurrentMap<SessionId, DateTime> datetime_cache(
```
I merely edited this

thomasalanlee commented on configuration/local/policy.yml:
```diff
@@ -14,14 +14,16 @@ logging:
-  type: standalone
-  expiration: 8h
-  persistenceToFileEnabled: false
+cache:
```
A general point but I'm not sure we should call infinispan a cache. We don't use it to increase access speed to a backing store so its more just a 'store' or a 'repository'

adityapahuja replied:
I see `cache`  has been replaced with `sessionStoreConfiguration`.  👍

thomasalanlee commented on hub/policy/src/main/java/uk/gov/ida/hub/policy/factories/CacheFactory.java:
```diff
@@ -0,0 +1,27 @@
+import java.util.concurrent.ConcurrentMap;
+
+public class CacheFactory {
+
```
We could refactor this so the getStateCache() and getDateTimeCache becomes a single method get(String cachename). We could also replace the strings with an enum to prevent invalid cache names being passed.

andy-paine commented on hub/policy/src/main/java/uk/gov/ida/hub/policy/PolicyConfiguration.java:
```diff
@@ -157,4 +158,9 @@ public boolean isEidasEnabled() {
     }
+
+    @Override
+    public InfinispanConfiguration getInfinispan() {
```
This is implementing `getInfinispan()` from the `InfinispanServiceConfiguration` interface (which is from one of our libs) so I don't really have any control over the naming 😞See [dropwizard-infinispan](https://github.com/alphagov/dropwizard-infinispan/blob/master/src/main/java/uk/gov/ida/shared/dropwizard/infinispan/config/InfinispanServiceConfiguration.java) for the poor naming

willp-bl replied:
fair enough - could be a good change to make in the lib if possible

andy-paine commented:
@willp-bl @thomasalanlee I've updated based on suggestions.
Points of note:
* Changed (generally) from cache -> store however all the infinispan lib stuff still refers to it as cache
* Named the two concurrent maps to `SessionExpirationStore` and `SessionStateStore`
* Added some braces @willp-bl 😉- no more [goto fail](https://nakedsecurity.sophos.com/2014/02/24/anatomy-of-a-goto-fail-apples-ssl-bug-explained-plus-an-unofficial-patch/)
* Extracted the method in the `SessionStoreFactory` (previously `CacheFactory`)

andy-paine commented:
I've added a new commit related to metrics/healthchecking the 2 session stores. Noteworthy points:
- I only bothered to do the State store as the datetime store is hopefully going away
- This _does_ log session IDs in the healthcheck but I don't consider these sensitive as it is on the admin endpoint and these things are already logged
- I don't really understand Guice + DropWizard all that well so what I've done with the bundle stuff might be absolutely horrendous but it works on my machine
- Our integration tests for Policy are worringly brittle due to using stuff like `PolicyIntegrationApplication`

When we actually run this, we should be able to see a nice graph of the `transition` cache slowly approaching the `existing` cache count before exactly tracking each other.

willp-bl commented:
I don't think that the figures will ever exactly match due to the expiry times being handled separately by the respective stores/caches

adityapahuja commented on hub/policy/src/test/java/uk/gov/ida/hub/policy/controllogic/TransitionStoreTest.java:
```diff
@@ -0,0 +1,408 @@
+        TransitionStore transitionStore = new TransitionStore(existing, Optional.of(transition));
+        transitionStore.putAll(putAll);
+        assertThat(existing.containsKey(key2) && existing.containsKey(key3)).isTrue();
+        assertThat(existing.containsKey(key2) && existing.containsKey(key3)).isTrue();
```
Can this line be `assertThat(transition.containsKey(key2) && existing.containsKey(key3)).isTrue();`?

adityapahuja commented on hub/policy/src/test/java/uk/gov/ida/hub/policy/controllogic/TransitionStoreTest.java:
```diff
@@ -0,0 +1,408 @@
+        transitionStore.putAll(putAll);
+        assertThat(existing.containsKey(key2) && existing.containsKey(key3)).isTrue();
+        assertThat(existing.containsKey(key2) && existing.containsKey(key3)).isTrue();
+    }
```
Can this line be assertThat(transition.containsKey(key2) && transition.containsKey(key3)).isTrue();?

adityapahuja commented on hub/policy/src/test/java/uk/gov/ida/hub/policy/controllogic/TransitionStoreTest.java:
```diff
@@ -0,0 +1,510 @@
+
+    @Test
+    public void getDifferingKeysShouldGiveSessionIdsOnlyInExistingMap() {
+        SessionId key1 = new SessionId("key1");
```
getDifferingSessionIds() will return empty if both existing and transition have same session ids but these session ids have different values. Can we add a test case to test this scenario?

adityapahuja commented on hub/policy/src/main/java/uk/gov/ida/hub/policy/factories/SessionStoreFactory.java:
```diff
@@ -0,0 +1,32 @@
+
+    public static ConcurrentMap<SessionId, DateTime> getSessionExpirationStore(
+            InfinispanCacheManager infinispanCacheManager
+    ) {
```
Can we have a more general method `getSessionStore(InfinispanCacheManager, CacheName)` instead of having two specialised methods `getSessionStateStore(InfinispanCacheManager)` and `getSessionExpirationStore(InfinispanCacheManager)` and replace hardcoded values `state_cache` and `datetime_cache` with CacheName Enum values.

adityapahuja commented on hub/policy/src/main/java/uk/gov/ida/hub/policy/factories/SessionStoreFactory.java:
```diff
@@ -0,0 +1,32 @@
+        return getSessionStore(infinispanCacheManager, "state_cache");
+    }
+
+    public static ConcurrentMap<SessionId, DateTime> getSessionExpirationStore(
```
getSessionExpirationStore() method is not used anywhere in policy. Is the method needed in the future?

adityapahuja commented on hub/policy/src/main/java/uk/gov/ida/hub/policy/PolicyApplication.java:
```diff
@@ -66,11 +67,19 @@ public final void initialize(Bootstrap<PolicyConfiguration> bootstrap) {
-                .modules(getPolicyModule(), new EventEmitterModule(),  bindInfinispan(infinispanBundle.getInfinispanCacheManagerProvider()))
+                .modules(getPolicyModule(),
+                        new EventEmitterModule(),
+                        getSessionModule(infinispanBundle.getInfinispanCacheManagerProvider(),
```
Can we replace this line with `new SessionModule(infinispanBundle.getInfinispanCacheManagerProvider(), sessionBundle.getSessionStateStoreProvider())` and remove `getSessionModule()` method from this class? Less code -> easy to read.

adityapahuja commented on hub/policy/src/main/java/uk/gov/ida/hub/policy/configuration/InfinispanConfiguration.java:
```diff
@@ -0,0 +1,24 @@
+    @JsonProperty
+    private uk.gov.ida.shared.dropwizard.infinispan.config.InfinispanConfiguration configuration;
+
+    public boolean isEnabled() {
```
`isEnabled()` method is used in `StateControllerFactoryTest.java` only. Is the method needed in the future?

adityapahuja commented:
There are some minor changes. All other changes look good.

andy-paine commented:
This is probably of a lot lower value if we are going to have to fail over to a new platform anyway. If we still want the concept it is still easy enough to pull out and the `SessionStore` stuff in the Redis PR I just raised is far simpler anyway so it would be quicker to just re-do this.
